commons/src/main/java/io/aiven/kafka/tieredstorage/commons/transform/FetchChunkEnumeration.java

```diff
@@ -0,0 +1,147 @@
+     * @param manifest provides to index to build response from
+     * @param firstPosition original offset range position to start from
+     * @param lastPosition original offset range position to finish
+     */
+    public FetchChunkEnumeration(final ChunkManager chunkManager,
+                                 final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                 final SegmentManifest manifest,
+                                 final int firstPosition,
+                                 final int lastPosition) {
+        this.chunkManager = Objects.requireNonNull(chunkManager, "chunkManager cannot be null");
+        this.remoteLogSegmentMetadata =
+            Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata cannot be null");
+        this.manifest = Objects.requireNonNull(manifest, "manifest cannot be null");
+        this.firstPosition = firstPosition;
+        this.lastPosition = lastPosition;
+
+        this.chunkIndex = manifest.chunkIndex();
+
+        final Chunk firstChunk = getFirstChunk(firstPosition);
+        startChunkId = firstChunk.id;
+        currentChunkId = startChunkId;
+        final Chunk lastChunk = getLastChunk(lastPosition);
+        lastChunkId = lastChunk.id;
+    }
+
+    private Chunk getFirstChunk(final int fromPosition) {
+        final Chunk firstChunk = chunkIndex.findChunkForOriginalOffset(fromPosition);
+        if (firstChunk == null) {
+            throw new IllegalArgumentException("Invalid start position "
+                + fromPosition + " in segment " + remoteLogSegmentMetadata);
+        }
+        return firstChunk;
+    }
+
+    private Chunk getLastChunk(final int endPosition) {
+        final Chunk chunkForOriginalOffset = chunkIndex.findChunkForOriginalOffset(endPosition);
+        if (chunkForOriginalOffset == null) {
+            final List<Chunk> chunks = chunkIndex.chunks();
+            return chunks.get(chunks.size() - 1);
+        } else {
+            return chunkForOriginalOffset;
+        }
+    }
+
+    @Override
+    public boolean hasMoreElements() {
+        return currentChunkId <= lastChunkId;
+    }
+
+    @Override
+    public InputStream nextElement() {
+        if (!hasMoreElements()) {
+            throw new NoSuchElementException();
+        }
+
+        InputStream chunkContent = getChunkContent(currentChunkId);
+
+        final Chunk currentChunk = chunkIndex.chunks().get(currentChunkId);
+        final int chunkStartPosition = currentChunk.originalPosition;
+        final boolean isAtFirstChunk = currentChunkId == startChunkId;
+        final boolean isAtLastChunk = currentChunkId == lastChunkId;
+        final boolean isSingleChunk = isAtFirstChunk && isAtLastChunk;
+        if (isSingleChunk) {
+            final int toSkip = firstPosition - chunkStartPosition;
+            try {
+                chunkContent.skip(toSkip);
+                final int chunkSize = lastPosition - chunkStartPosition - toSkip;
```

Comment:
This is basically `lastPosition - firstPosition` and IMO having it like this will be a little cleaner

Reply:
Agree, I like it better. Pushing changes now.